httomo/runner/loader.py

```diff
@@ -151,6 +151,97 @@ class UserDefinedAngles(NamedTuple):
 
 AnglesConfig: TypeAlias = Union[RawAngles, UserDefinedAngles]
 
+class PreviewDimConfig(NamedTuple):
+    start: int
+    stop: int
+
+class PreviewConfig(NamedTuple):
+    angles: PreviewDimConfig
+    detector_y: PreviewDimConfig
+    detector_x: PreviewDimConfig
+
+class Preview:
+    def __init__(
+        self,
+        preview_config: PreviewConfig,
+        dataset: h5py.Dataset,
+        image_key: Optional[h5py.Dataset],
+    ) -> None:
+        self.config = preview_config
+        self._dataset = dataset
+        self._image_key = image_key
+        self._check_within_data_bounds()
+        self._data_indices: Optional[List[int]] = None
+        self._global_shape: Optional[Tuple[int, int, int]] = None
+
+    def _check_within_data_bounds(self) -> None:
+        shape = self._dataset.shape
+        if self.config.angles.stop > shape[0]:
+            raise ValueError(
+                f"Preview indices in angles dim exceed bounds of data: "
+                f"start={self.config.angles.start}, stop={self.config.angles.stop}"
+            )
+
+        if self.config.detector_y.stop > shape[1]:
+            raise ValueError(
+                f"Preview indices in det y dim exceed bounds of data: "
+                f"start={self.config.detector_y.start}, "
+                f"stop={self.config.detector_y.stop}"
+            )
+
+        if self.config.detector_x.stop > shape[2]:
+            raise ValueError(
+                f"Preview indices in det x dim exceed bounds of data: "
+                f"start={self.config.detector_x.start}, "
+                f"stop={self.config.detector_x.stop}"
+            )
+
+    def _calculate_data_indices(self) -> List[int]:
+        if self._image_key is not None:
+            indices = np.where(
+                self._image_key[:] == 0
+            )[0].tolist()
+        else:
+            no_of_angles = self._dataset.shape[0]
+            indices = list(range(no_of_angles))
+
+        preview_data_indices = np.arange(
+            self.config.angles.start, self.config.angles.stop
+        )
+
+        intersection = np.intersect1d(indices, preview_data_indices)
+        if not np.array_equal(preview_data_indices, intersection):
+            self.config = PreviewConfig(
+                angles=PreviewDimConfig(start=intersection[0], stop=intersection[-1] + 1),
+                detector_y=self.config.detector_y,
+                detector_x=self.config.detector_x,
+            )
+
+        return intersection.tolist()
+
+    def _calculate_global_shape(self) -> Tuple[int, int, int]:
+        if self._data_indices is None:
+            self._data_indices = self._calculate_data_indices()
+
+        return (
+            len(self._data_indices),
+            len(range(self.config.detector_y.start, self.config.detector_y.stop)),
+            len(range(self.config.detector_x.start, self.config.detector_x.stop)),
```

Comment:
```suggestion
            self.config.detector_y.stop - self.config.detector_y.start,
            self.config.detector_x.stop - self.config.detector_x.start,
```

Reply:
Thanks for noting this, the brain does unnecessarily complicated things sometimes! :laughing: Done in [0ebdedb](https://github.com/DiamondLightSource/httomo/pull/209/commits/0ebdedb11c9eac3b72e83ddadfe1d055b66caf11)